File: 09-backtracking/00051-N_Queens/00051-n_queens.cpp
// 51: N Queens
// https://leetcode.com/problems/n-queens


#include <iostream>
#include <vector>

using namespace std;

class Solution {
private:
    void backtrack(int n, int row, vector<string>& nQueens, vector<vector<string>>& result) {
        if (row == n) {
            result.push_back(nQueens);
            return;
        }

        for (int col=0; col<n; col++) {
            if (isValid(nQueens, row, col, n)) {
                nQueens[row][col] = 'Q';
                backtrack(n, row+1, nQueens, result);
                nQueens[row][col] = '.';
            }
        }
    }

    bool isValid(std::vector<std::string> &nQueens, int row, int col, int &n) {
        for (int i=0; i<row; i++)
            if (nQueens[i][col] == 'Q')
                return false;

        for (int i=row-1, j=col-1; i>=0 && j>=0; i--, j--)
            if (nQueens[i][j] == 'Q')
                return false;

        for (int i=row-1, j=col+1; i>=0 && j<n; i--, j++)
            if (nQueens[i][j] == 'Q')
                return false;

        return true;
    }

public:
    // SOLUTOIN
    vector<vector<string>> solveNQueens(int n) {
        vector<string> nQueens(n, string(n, '.'));
        vector<vector<string>> result;
        backtrack(n, 0, nQueens, result);
        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    int n = 4;

    // OUTPUT
    auto result = o.solveNQueens(n);

    cout << "["; for (auto x : result) {
        cout << "["; for (auto y : x) {
            cout << "\"" << y << "\", ";
        } cout << "\b\b], ";
    } cout << "\b\b]" << endl;

    return 0;
}

File: 09-backtracking/00051-N_Queens/00051-n_queens.java
// 51: N Queens
// https://leetcode.com/problems/n-queens


import java.util.List;
import java.util.ArrayList;

class Solution {
    private void backtrack(int n, int row, List<String> nQueens, List<List<String>> result) {
        if (row == n) {
            result.add(new ArrayList<>(nQueens));
            return;
        }

        for (int col=0; col<n; col++) {
            if (isValid(nQueens, row, col, n)) {
                StringBuilder r = new StringBuilder(nQueens.get(row));
                r.setCharAt(col, 'Q');
                nQueens.set(row, r.toString());
                backtrack(n, row + 1, nQueens, result);
                r.setCharAt(col, '.');
                nQueens.set(row, r.toString());
            }
        }
    }

    private boolean isValid(List<String> nQueens, int row, int col, int n) {
        for (int i=0; i<row; i++)
            if (nQueens.get(i).charAt(col) == 'Q')
                return false;

        for (int i=row-1, j=col-1; i>=0 && j>=0; i--, j--)
            if (nQueens.get(i).charAt(j) == 'Q')
                return false;

        for (int i=row-1, j=col+1; i>=0 && j<n; i--, j++)
            if (nQueens.get(i).charAt(j) == 'Q')
                return false;

        return true;
    }

    
    // SOLUTION
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> result = new ArrayList<>();
        List<String> nQueens = new ArrayList<String>();
        for (int i=0; i<n; i++) {
            nQueens.add(".".repeat(n));
        }
        backtrack(n, 0, nQueens, result);
        return result;
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int n = 4;

        // OUTPUT
        var result = o.solveNQueens(n);
        System.out.println(result);
    }
}
